### DISCUSSION TITLE
[RRFC] platform-specific configuration values

### DISCUSSION THREAD

**Discussion by isaacs** [npm/rfcs#60] [General]
Suggested in https://github.com/npm/rfcs/pull/36 

Request for an RFC to explore having platform-specific config values, which is particularly relevant if configuration values would ride along with installed packages as implied by https://github.com/npm/rfcs/issues/59.

Implementation is relatively straightforward, but the following bikesheds need to be addressed:

- Fields we wish to filter on.  Is it _just_ `process.platform`, or are there others?
- Is a match/not-match syntax sufficient?  So a `platform: win32` section would provide config overrides for windows, and `platform: !win32` would provide config overrides for non-windows.  Would we need version comparisons, or other things if we're filtering based on node version etc?
- What happens if multiple configs match the current platform?  If you set `platform: !win32` and then later `platform: darwin`, do we prioritize the most specific, or is it order-dependent?  (Note that order dependency is particularly fraught when dealing with JS objects, where key order can be munged without realizing it.)

Straw proposal to go from:

```ini
; in an ini-formatted .npmrc file
script-shell = /bin/sh

[platform: win32]
script-shell = c:\Program Files\git\bin\bash.exe

[platform: linux]
script-shell = /bin/bash
```

```json
{
  "name": "package-json-with-configs",
  "npm": {
    "script-shell": "/bin/sh",
    "platform: win32": {
      "script-shell": "c:\\Program Files\\git\\bin\\bash.exe"
    },
    "platform: linux": {
      "script-shell": "/bin/bash"
    }
}
```

**Comment by mikemimik** [comment]
In the context of there being "platform specific" **config** _(generally speaking)_, then #28 seems to be tangentially related.
